ROX-26411: Improve the upgrader nudge texts

vikin91 opened the pull request:
### Description

This PR updates some texts regarding the upgradability of the secured clusters. 

It addresses the following issues:
1. We should not indicate that a version is managed by Helm if it is managed by an operator
2. We should not speak of **upgrade** when in fact we mean **downgrade**
3. We should not use the word **required** when we propose a **downgrade**

in the following manner:
1. Instead we indicate that the version is not managed by ACS but by external tools - for example Helm or an operator
2. We use **upgrade** only if sensor is behind Central. In opposite case, we use word **downgrade**
3. If we discover that sensor can be downgraded, we do not use **required** but **possible** instead. The goal is to not stress the necessity.

What this PR is not addressing:
1. Instead of proposing a Sensor downgrade on **self-managed** cluster, we should recommend an upgrade of Central
2. Instead of proposing a Sensor downgrade on **managed** cluster, we should recommend **waiting** for an upgrade of Central

### User-facing documentation

- [x] CHANGELOG is updated **OR** update is not needed
- [x] [documentation PR](https://spaces.redhat.com/display/StackRox/Submitting+a+User+Documentation+Pull+Request) is created and is linked above **OR** is not needed

### Testing and quality

<!--
General Availability requirements: https://github.com/stackrox/stackrox/blob/master/PR_GA.md
Feature Flags usage: https://github.com/stackrox/stackrox/blob/master/pkg/features/README.md
-->

- [x] the change is production ready: the change is GA or otherwise the functionality is gated by a feature flag
- [x] CI results are inspected

#### Automated testing

<!--
If no tests have been contributed, please explain why unless it's obvious,
e.g., the PR is a one-line comment change.
-->

- [ ] added unit tests
- [ ] added e2e tests
- [ ] added regression tests
- [ ] added compatibility tests
- [x] modified existing tests

#### How I validated my change

##### Who manages the cluster?

1. Deployed to GKE using Helm
2. Manually changed sensor version to lower
3. Expected the cluster page to report that the cluster version is not managed by ACS
4. Got: The clusters page gave a link to upgrade the cluster. Apparently the discovery of installation method does not work well. I [asked the install team
](https://redhat-internal.slack.com/archives/C073B14UE10/p1727696319624519) and moved on (this is outside of the scope for this PR).
5. I manually edited the secret `helm-cluster-config` and changed the respective values to force this cluster being recognized as managed by helm: `managedBy: MANAGER_TYPE_HELM` and `notHelmManaged: false`. After that, I restarted sensor.

Clusters table view:
![Screenshot 2024-10-08 at 17 30 05](https://github.com/user-attachments/assets/4c981ea2-1f2a-4e2f-bb92-532a5242a6e9)


Cluster details view:
![Screenshot 2024-10-08 at 17 22 16](https://github.com/user-attachments/assets/e528660f-3e38-42e3-acab-4a83040dfd93)


##### Downgrade situation

To validate the downgrade scenario, I would require (1) a sensor that has higher version, and (2) central that is based on this branch. I built two custom images from this branch - one 4.6.99, the other 4.6.98 - and shuffled them to trigger the downgrade scenario. The result is as on the screenshot.

![Screenshot 2024-09-30 at 15 08 30](https://github.com/user-attachments/assets/31d80e86-980f-44ec-94dc-505a7c829d77)

Cluster details view:
![Screenshot 2024-09-30 at 15 10 01](https://github.com/user-attachments/assets/cde34932-a5a4-4763-9713-9a9f564c6f70)

pedrottimark commented on ui/apps/platform/src/Containers/Clusters/cluster.helpers.tsx:
```diff
@@ -187,13 +187,17 @@ const upgradeStates: UpgradeStates = {
     },
     MANUAL_UPGRADE_REQUIRED: {
-        displayValue: 'Manual upgrade required',
+        displayValue: 'Version not managed by ACS',
```
Piotr, how does **not managed by ACS** here and in corresponding backend messages compare to **managed by external tools**?

The reason I ask is that even if we replace **ACS** with **RHACS** for consistency with elsewhere in UI, we avoid proper name unless necessary for meaning, because of product branding for downstream versus upstream.

vikin91 replied:
I updated the PR description with the screenshots - please have a look.

vikin91 commented on central/sensor/service/connection/upgradecontroller/trigger.go:
```diff
@@ -129,11 +129,11 @@ func (u *upgradeController) doTriggerUpgrade() (common.MessageInjector, *central
 		// We still allow upgrade triggers in this case.
 	case storage.ClusterUpgradeStatus_MANUAL_UPGRADE_REQUIRED:
-		return nil, nil, errors.Errorf("manual upgrade required for cluster %s; cannot trigger upgrade", u.clusterID)
+		return nil, nil, errors.Errorf("ACS does not manage version for secured cluster %s; cannot trigger upgrade/downgrade", u.clusterID)
```
```suggestion
		return nil, nil, errors.Errorf("RHACS does not manage version for secured cluster %s; cannot trigger upgrade/downgrade", u.clusterID)
```

vikin91 commented on ui/apps/platform/src/Containers/Clusters/cluster.helpers.test.js:
```diff
@@ -338,14 +338,14 @@ describe('cluster helpers', () => {
         });
 
-        it('should return "Manual upgrade required" if upgradeStatus -> upgradability is MANUAL_UPGRADE_REQUIRED', () => {
+        it('should return "Version not managed by ACS" if upgradeStatus -> upgradability is MANUAL_UPGRADE_REQUIRED', () => {
```
```suggestion
        it('should return "Version not managed by RHACS" if upgradeStatus -> upgradability is MANUAL_UPGRADE_REQUIRED', () => {
```

vikin91 commented on ui/apps/platform/src/Containers/Clusters/cluster.helpers.test.js:
```diff
@@ -338,14 +338,14 @@ describe('cluster helpers', () => {
             const received = findUpgradeState(testUpgradeStatus);
 
-            const expected = { displayValue: 'Manual upgrade required', type: 'intervention' };
+            const expected = { displayValue: 'Version not managed by ACS', type: 'intervention' };
```
```suggestion
            const expected = { displayValue: 'Version not managed by RHACS', type: 'intervention' };
```

vikin91 commented on ui/apps/platform/src/Containers/Clusters/cluster.helpers.tsx:
```diff
@@ -187,13 +187,17 @@ const upgradeStates: UpgradeStates = {
     },
     MANUAL_UPGRADE_REQUIRED: {
-        displayValue: 'Manual upgrade required',
+        displayValue: 'Version not managed by ACS',
```
```suggestion
        displayValue: 'Version not managed by RHACS',
```

vikin91 reviewed (commented):
Self-review.

jschnath reviewed (approved):
As seen in the demo today, looks good to me 👍

vikin91 commented:
/test gke-sensor-integration-tests

alwayshooin commented:
do we need to update documentation for this? cc: @kcarmichael08

gaurav-nelson commented on central/sensor/service/connection/connection_impl.go:
```diff
@@ -830,7 +830,7 @@ func (c *sensorConnection) ObjectsDeletedByReconciliation() (map[string]int, boo
 func (c *sensorConnection) CheckAutoUpgradeSupport() error {
 	if c.sensorHello.GetHelmManagedConfigInit() != nil && !c.sensorHello.GetHelmManagedConfigInit().GetNotHelmManaged() {
-		return errors.New("cluster is Helm-managed and does not support auto upgrades; use 'helm upgrade' or a Helm-aware CD pipeline for upgrades")
+		return errors.New("secured cluster version is managed by external tools (helm, operator)")
```
```suggestion
		return errors.New("External tools (Helm or Operator) control the secured cluster version.")
```

vikin91 replied:
This message reads good, however our linter may not be happy about that, as error messages in Go should not start with capital letter and not end with a period. I will check whether we can add some exception here.

gaurav-nelson commented on central/sensor/service/connection/upgradecontroller/trigger.go:
```diff
@@ -129,11 +129,11 @@ func (u *upgradeController) doTriggerUpgrade() (common.MessageInjector, *central
 		// We still allow upgrade triggers in this case.
 	case storage.ClusterUpgradeStatus_MANUAL_UPGRADE_REQUIRED:
-		return nil, nil, errors.Errorf("manual upgrade required for cluster %s; cannot trigger upgrade", u.clusterID)
+		return nil, nil, errors.Errorf("RHACS does not manage version for secured cluster %s; cannot trigger upgrade/downgrade", u.clusterID)
```
```suggestion
		return nil, nil, errors.Errorf("RHACS does not control the version for the secured cluster %s; cannot trigger upgrade or downgrade", u.clusterID)
```

vikin91 replied:
Addressed in [f81eac1](https://github.com/stackrox/stackrox/pull/12833/commits/f81eac1c3bb3705d1a08dd84fbc5e019187b768f)

gaurav-nelson commented on central/sensor/service/connection/upgradecontroller/trigger.go:
```diff
@@ -129,11 +129,11 @@ func (u *upgradeController) doTriggerUpgrade() (common.MessageInjector, *central
 		return nil, nil, errors.Errorf("sensor for cluster %s is already up-to-date; cannot trigger upgrade", u.clusterID)
 	default:
-		return nil, nil, errors.Errorf("unknown upgradability status of sensor for cluster %s; cannot trigger upgrades", u.clusterID)
+		return nil, nil, errors.Errorf("unknown upgradability status of sensor for cluster %s; cannot trigger upgrade/downgrade", u.clusterID)
```
```suggestion
		return nil, nil, errors.Errorf("The sensor's upgradability status for cluster %s is unclear; cannot trigger upgrade or downgrade", u.clusterID)
```

vikin91 replied:
The message is actually for the secured cluster. Sensor was used as a name, because we used to call secured clusters like this. Maybe I would change it to secured cluster?

vikin91 replied:
Addressed in [f81eac1](https://github.com/stackrox/stackrox/pull/12833/commits/f81eac1c3bb3705d1a08dd84fbc5e019187b768f) - I changed sensor to secured cluster and rephrased a bit.

gaurav-nelson commented on ui/apps/platform/src/Containers/Clusters/cluster.helpers.test.js:
```diff
@@ -338,14 +340,31 @@ describe('cluster helpers', () => {
         });
 
-        it('should return "Manual upgrade required" if upgradeStatus -> upgradability is MANUAL_UPGRADE_REQUIRED', () => {
+        it(`should return "Version not managed by ${getProductBranding().reportName}" if upgradeStatus -> upgradability is MANUAL_UPGRADE_REQUIRED`, () => {
```
If we know this message is for Sensor only, maybe we can say `Sensor is not managed by ${getProductBranding().reportName}.`

gaurav-nelson commented on ui/apps/platform/src/Containers/Clusters/cluster.helpers.tsx:
```diff
@@ -187,13 +188,17 @@ const upgradeStates: UpgradeStates = {
     },
     MANUAL_UPGRADE_REQUIRED: {
-        displayValue: 'Manual upgrade required',
+        displayValue: `Version not managed by ${getProductBranding().reportName}`,
```
Maybe `Sensor is not managed by ${getProductBranding().reportName}.` if we know this will only appear for sensor.

vikin91 commented:
I think, I addressed all the feedback. It is pretty tricky to generate new screenshots (I need to do a lot of magic to simulate different combinations of versions and it takes ~1hr) - the currently included screenshots contain only some changes. I will try to generate new ones if that is necessary, but it would be cool to review based on the code diff. Thanks!

vikin91 commented:
/retest

vikin91 commented:
/retest

vikin91 commented:
@alwayshooin @pedrottimark any chance getting this a UI review? It is a small PR so I hope we could merge it soon ;)

gaurav-nelson reviewed (approved):
LGTM

pedrottimark reviewed (commented):
Thank you for contributing. However, there is a logical contradiction here:

> This message is shown for a cluster in the clusters page, so the context is there and the message should be short.

1. External tools (Helm or Operator) control the secured cluster version.
2. Secured cluster version is not managed by Red Hat Advanced Cluster Security (RHACS).

The first is 14 characters longer than the second.

vikin91 commented:
> Thank you for contributing. However, there is a logical contradiction here:
> 
> > This message is shown for a cluster in the clusters page, so the context is there and the message should be short.
> 
> 1. External tools (Helm or Operator) control the secured cluster version.
> 2. Secured cluster version is not managed by Red Hat Advanced Cluster Security (RHACS).
> 
> The first is 14 characters longer than the second.

@pedrottimark , the discussion about short message was between 
1. `Version not managed by ACS` versus 
2. `Secured cluster version is not managed by ${getProductBranding().reportName}.`. 

I would personally prefer the shorter version of that message, but I am not sure whether this message must be short - maybe that is a requirement that I artificially created? I would actually appreciate your comment on that. If that message can be longer, then let's pick the second variant as it is more accurate.
By the way, shouldn't we use `RHACS` for short instead of the full `Red Hat Advanced Cluster Security (RHACS)`?

The message `External tools (Helm or Operator) control the secured cluster version.` is displayed in a tooltip and in the cluster details page. In both of those places we have enough space to display it nicely.

It is the EOD for me, so I will check tomorrow if there are any comments and if not, I will merge this. In case we have second thoughts on the length of the texts, I can open a follow-up PR.

alwayshooin commented:
> I would personally prefer the shorter version of that message, but I am not sure whether this message must be short - maybe that is a requirement that I artificially created? I would actually appreciate your comment on that. If that message can be longer, then let's pick the second variant as it is more accurate. By the way, shouldn't we use `RHACS` for short instead of the full `Red Hat Advanced Cluster Security (RHACS)`?

thanks piotr, i think using `RHACS` makes sense if that works for @gaurav-nelson ? i'll add a commit to add the short name version
